### PR TITLE
Allow disabling navigation items via internal feature flags

### PR DESCRIPTION
**Pull request by ari-aviran**
## Summary

Part of https://github.com/elastic/security-team/issues/2595

This PR adds "internal feature flags" functionality for hiding in-development features via a code constant, and also adds a mechanism to hide navigation items from our left-hand-side navigation bar.

### CODE REVIEW

File: x-pack/plugins/cloud_security_posture/public/common/navigation/constants.ts [hunk 1]
```diff
@@ -7,14 +7,24 @@
...
 import * as TEXT from './translations';
 import type { CspNavigationItem } from './types';
+import { INTERNAL_FEATURE_FLAGS } from '../../../common/constants';
 
 // This makes sure our routes are constrained by CspNavigationItem
 // By doing this we get to do a type-check on `v` but keep the readonly properties
 const createRoutes = <T extends readonly CspNavigationItem[]>(routes: T) =>
-  Object.fromEntries(routes.map((v) => [v.id, v])) as { [Page in T[number] as Page['id']]: Page };
+  Object.fromEntries(routes.map((v) => [v.id, v])) as {
+    [Page in T[number] as Page['id']]: CspNavigationItem;
```
Comment (ari-aviran) [inline x-pack/plugins/cloud_security_posture/public/common/navigation/constants.ts:16]: Changed the type of the values in the mapping to `CspNavigationItem` instead of an inferred literal type. The reason is that I added an optional property to `CspNavigtionItem` and since consumers of these values iterate over the values it requires a consistent type (instead of a union type of all possible literal values)

File: x-pack/plugins/cloud_security_posture/public/common/navigation/constants.ts [hunk 2]
```diff
@@ -7,14 +7,24 @@
...
+  };
 
 // This is the source-of-truth of the routes. links/types are derived
-export const allNavigationItems = createRoutes([
-  { name: TEXT.DASHBOARD, path: '/dashboard', id: 'dashboard' },
-  { name: TEXT.FINDINGS, path: '/findings', id: 'findings' },
-] as const);
+export const getAllNavigationItems = () =>
```
Comment (ari-aviran) [inline x-pack/plugins/cloud_security_posture/public/common/navigation/constants.ts:20]: Made this a function instead of an object to allow mocking in tests

File: x-pack/plugins/cloud_security_posture/public/components/page_template.tsx [hunk 3]
```diff
@@ -11,34 +11,36 @@ import {
...
-  allNavigationItems
-).map((route) => ({
-  ...route,
-  renderItem: () => (
-    <NavLink to={route.path} activeStyle={activeItemStyle}>
-      {route.name}
-    </NavLink>
-  ),
-}));
+const getSideNavItems = (): NonNullable<KibanaPageTemplateProps['solutionNav']>['items'] =>
```
Comment (ari-aviran) [inline x-pack/plugins/cloud_security_posture/public/components/page_template.tsx:19]: This is now lazily evaluated (during render instead of on "import") so we can setup mocks in tests before this is evaluated

File: x-pack/plugins/cloud_security_posture/public/common/navigation/constants.ts (hunk 1 above)
Comment (orouz) [reply x-pack/plugins/cloud_security_posture/public/common/navigation/constants.ts:16]: `Page` is constrained by `CspNavigationItem` 

i don't see how this change helps. if any, we now have less type information. 

can you give an example of why this was needed? what didn't work for you when iterating?

Comment (ari-aviran) [reply x-pack/plugins/cloud_security_posture/public/common/navigation/constants.ts:16]: This is indeed less "type information" but being overly specific with literal types can cause issues, for example:
image

File: x-pack/plugins/cloud_security_posture/public/common/navigation/types.ts [hunk 4]
```diff
@@ -6,7 +6,8 @@
  */
 
 export interface CspNavigationItem {
-  name: string;
-  path: string;
-  id: string;
+  readonly name: string;
+  readonly path: string;
+  readonly id: string;
+  readonly disabled?: boolean;
```
Comment (orouz) [inline x-pack/plugins/cloud_security_posture/public/common/navigation/types.ts:12]: consider not adding the `Benchmark` page to our routes if it is disabled.

File: x-pack/plugins/cloud_security_posture/public/common/navigation/constants.ts (hunk 1 above)
Comment (orouz) [reply x-pack/plugins/cloud_security_posture/public/common/navigation/constants.ts:16]: you could still fix this easily without losing type information.

1 - dont add benchmark to array if not used 
2 - add `disabled: false` to other pages so `disabled` exists (removing the optional modifier)
3 - adding a default `disabled: false` on the type level if `Page` doesn't have one.

File: x-pack/plugins/cloud_security_posture/public/components/page_template.test.tsx [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+    const navigationItemName = chance.word();
+    const navigationItem = createNavigationItemFixture({
+      name: navigationItemName,
+      disabled: true,
+    });
+    (getAllNavigationItems as jest.Mock).mockReturnValue({ ...navigationItem });
+
+    renderCspPageTemplate();
+
+    expect(screen.queryByText(navigationItemName)).not.toBeInTheDocument();
```
Comment (orouz) [inline x-pack/plugins/cloud_security_posture/public/components/page_template.test.tsx:51]: is `UnknownRoute` shown when it is `disabled`? if so, maybe assert that too

File: x-pack/plugins/cloud_security_posture/public/pages/findings/findings.test.tsx [hunk 6]
```diff
@@ -25,12 +24,9 @@ beforeEach(() => {
 });
 
 const FindingsComponentWithTestProvider = () => {
-  const core = coreMock.createStart();
-  const params = coreMock.createAppMountParameters();
   const dataMock = dataPluginMock.createStartContract();
-  const services = { core, deps: { data: dataMock }, params };
   return (
-    <TestProvider {...services}>
```
Comment (orouz) [inline x-pack/plugins/cloud_security_posture/public/pages/findings/findings.test.tsx:33]: can you clarify why were these removed?

File: x-pack/plugins/cloud_security_posture/public/pages/findings/findings.tsx [hunk 7]
```diff
@@ -21,7 +21,7 @@ const pageHeader: EuiPageHeaderProps = {
 
 export const Findings = () => {
   const dataView = useKubebeatDataView();
-  useCspBreadcrumbs([allNavigationItems.findings]);
+  useCspBreadcrumbs([getAllNavigationItems().findings]);
```
Comment (orouz) [inline x-pack/plugins/cloud_security_posture/public/pages/findings/findings.tsx:24]: consider exporting a static object for our app and a function for testing (also not sure why we need a function for testing)

as now we call a function to get our static routes on every render, even though the data doesn't change.

File: x-pack/plugins/cloud_security_posture/public/components/page_template.tsx (hunk 3 above)
Comment (orouz) [inline x-pack/plugins/cloud_security_posture/public/components/page_template.tsx:19]: consider making this function take `CspNavigationItem` as it's argument
so we don't call `getAllNavigationItems` repeatedly

**Review by orouz — APPROVED** [review]
see comments

File: x-pack/plugins/cloud_security_posture/public/components/page_template.test.tsx (hunk 5 above)
Comment (ari-aviran) [reply x-pack/plugins/cloud_security_posture/public/components/page_template.test.tsx:51]: No, this is only checking the rendered navigation sidebar

File: x-pack/plugins/cloud_security_posture/public/pages/findings/findings.test.tsx (hunk 6 above)
Comment (ari-aviran) [reply x-pack/plugins/cloud_security_posture/public/pages/findings/findings.test.tsx:33]: I added these as defaults in `TestProvider`

File: x-pack/plugins/cloud_security_posture/public/components/page_template.test.tsx (hunk 5 above)
Comment (orouz) [reply x-pack/plugins/cloud_security_posture/public/components/page_template.test.tsx:51]: we might want to block rendering too

if someone goes to `/csp/benchmark` i think they'll see `hello world`

**Review by JordanSh — APPROVED** [review]
